pkg/controller/podautoscaler/horizontal.go

```diff
@@ -652,15 +689,96 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 	return a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
 }
 
+type NormalizationArg struct {
+	Key                 string
+	ScaleUpConstraint   autoscalingv2.HPAScaleConstraintValue
+	ScaleDownConstraint autoscalingv2.HPAScaleConstraintValue
+	MinReplicas         *int32
+	MaxReplicas         int32
+	CurrentReplicas     int32
+	DesiredReplicas     int32
+}
+
+// normalizeDesiredReplicas takes the metrics desired replicas value and normalizes it:
+// 1. It applies the basic conditions (i.e. < maxReplicas, > minReplicas, etc...)
+// 2. It applies the scale up/down limits from the hpaSpec.Constraints (i.e. add no more than 4 pods)
+// 3. It applies the constraints period (i.e. add no more than 4 pods per minute)
+// 4. It applies the constraints delay (i.e. add no more than 4 pods per minute, and pick the smallest recommendation during last 5 minutes)
+func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.HorizontalPodAutoscaler, key string, currentReplicas int32, prenormalizedDesiredReplicas int32) int32 {
+	normalizationArg := NormalizationArg{
+		Key:                 key,
+		ScaleUpConstraint:   generateHPAScaleUpConstraint(hpa.Spec.Constraints),
+		ScaleDownConstraint: generateHPAScaleDownConstraint(hpa.Spec.Constraints),
+		MinReplicas:         hpa.Spec.MinReplicas,
+		MaxReplicas:         hpa.Spec.MaxReplicas,
+		CurrentReplicas:     currentReplicas,
+		DesiredReplicas:     prenormalizedDesiredReplicas}
+	stabilizedRecommendation := a.stabilizeRecommendation(normalizationArg)
+	if stabilizedRecommendation != prenormalizedDesiredReplicas {
+		setCondition(hpa, autoscalingv2.AbleToScale, v1.ConditionTrue, "ScaleDownStabilized", "recent recommendations were higher than current one, applying the highest recent recommendation")
+	} else {
+		setCondition(hpa, autoscalingv2.AbleToScale, v1.ConditionTrue, "ReadyForNewScale", "recommended size matches current size")
+	}
+	desiredReplicas, condition, reason := a.convertDesiredReplicasWithConstraintRate(normalizationArg)
+	if desiredReplicas == stabilizedRecommendation {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionFalse, condition, reason)
+	} else {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, condition, reason)
+	}
+
+	return desiredReplicas
+}
+
+// getScaleUpChangePerPeriod returns number of pods added during the constrate rate period
+func getScaleUpChangePerPeriod(scaleUpEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
+	cutoff := time.Now().Add(-(time.Duration)(*args.ScaleUpConstraint.Rate.PeriodSeconds))
```

Comment:
> time.Now().Add(-(time.Duration)(*args.ScaleUpConstraint.Rate.PeriodSeconds))

Could we use `Sub(duration)` here instead of `Add(-duration)`?

Another option to make it easier to read and understand from the first sight would be to introduce a local variable (haven't been tested):

```go
period := time.Duration(*args.ScaleUpConstraint.Rate.PeriodSeconds)
cutoff := time.Now().Add(-period)
```

What do you think, does it worth it?

Reply:
`Sub(duration)` returns a `Duration`, not a `Time`. 

You example seems more readable to me.

Reply:
Yep, will do.